cmd/istioctl/mixer.go

```diff
@@ -0,0 +1,165 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"strings"
+
+	// TODO This should be github.com/googleapis/googleapis/google/rpc
+	rpc "github.com/googleapis/googleapis"
+
+	"github.com/ghodss/yaml"
+	"github.com/spf13/cobra"
+)
+
+// TODO This should come from something like istio.io/api instead of
+// being hand copied from istio.io/mixer.
+type mixerAPIResponse struct {
+	Data   interface{} `json:"data,omitempty"`
+	Status rpc.Status  `json:"status,omitempty"`
+}
+
+var (
+	mixerFile          string
+	mixerAPIServerAddr string
+
+	mixerCmd = &cobra.Command{
+		Use:          "mixer",
+		Short:        "Istio Mixer configuration",
+		SilenceUsage: true,
+		PersistentPreRunE: func(c *cobra.Command, args []string) error {
+			if mixerAPIServerAddr == "" {
+				return errors.New("no mixer API server specified (use --mixer)")
+			}
+			return nil
+		},
+	}
+
+	mixerRuleCmd = &cobra.Command{
+		Use:          "rule",
+		Short:        "Istio Mixer Rule configuration",
+		SilenceUsage: true,
+	}
+
+	mixerRuleCreateCmd = &cobra.Command{
+		Use:   "create <scope> <subject>",
+		Short: "Create Istio Mixer rules",
+		RunE: func(c *cobra.Command, args []string) error {
+			if len(args) != 2 || mixerFile == "" {
+				return errors.New(c.UsageString())
+			}
+			rule, err := ioutil.ReadFile(mixerFile)
+			if err != nil {
+				return fmt.Errorf("failed opening %s: %v", mixerFile, err)
+			}
+			return mixerRuleCreate(mixerAPIServerAddr, args[0], args[1], rule)
+		},
+	}
+	mixerRuleGetCmd = &cobra.Command{
+		Use:   "get <scope> <subject>",
+		Short: "Get Istio Mixer rules",
+		RunE: func(c *cobra.Command, args []string) error {
+			if len(args) != 2 {
+				return errors.New(c.UsageString())
+			}
+			out, err := mixerRuleGet(mixerAPIServerAddr, args[0], args[1])
+			if err != nil {
+				return err
+			}
+			fmt.Println(out)
+			return nil
+		},
+	}
+)
+
+func mixerRulePath(host, scope, subject string) string {
+	if !strings.HasPrefix(host, "http://") {
+		host = "http://" + host
+	}
+	return fmt.Sprintf("%s/api/v1/scopes/%s/subjects/%s/rules", host, scope, subject)
+}
+
+func mixerRuleCreate(host, scope, subject string, rule []byte) error {
+	request, err := http.NewRequest(http.MethodPut, mixerRulePath(host, scope, subject), bytes.NewReader(rule))
```

Comment:
Is there a better way to send a `put` request, or were you suggesting adding a cancellation context?  http.Client doesn't have a `Put` method.